packages/react-import-remote/src/hooks.ts

```diff
@@ -0,0 +1,123 @@
+  Initial = 'Initial',
+  Failed = 'Failed',
+  Complete = 'Complete',
+  Loading = 'Loading',
+}
+
+type Result<Imported = any> =
+  | {status: Status.Initial}
+  | {status: Status.Loading}
+  | {status: Status.Failed; error: Error}
+  | {status: Status.Complete; imported: Imported};
+
+export function useImportRemote<Imported = any>(
+  source: string,
+  options: Options<Imported>,
+): {
+  result: Result<Imported>;
+  intersectionRef: any;
+} {
+  const {defer = DeferTiming.Mount, nonce = '', getImport} = options;
+  const [result, setResult] = React.useState<Result>({status: Status.Initial});
+  const idleCallbackHandle = React.useRef<RequestIdleCallbackHandle | null>(
+    null,
+  );
+
+  const deferOption = React.useRef(defer);
+
+  if (deferOption.current !== defer) {
+    throw new Error(
+      [
+        'You’ve changed the defer strategy on an <ImportRemote />',
+        'component after it mounted. This is an expensive operation',
+        'that must be avoided in order to dismiss this error.',
+      ].join(' '),
+    );
+  }
+
+  let intersection;
+  let intersectionRef;
+
+  // Normally this would be dangerous but because we are
+  // guranteed to have thrown if the defer option changes
+  // we can be confident that a given use of this hook
+  // will only ever hit one of these two cases.
+  /* eslint-disable react-hooks/rules-of-hooks */
+  if (defer === DeferTiming.InViewport) {
+    [intersection, intersectionRef] = useIntersection();
+  }
+  /* eslint-enable react-hooks/rules-of-hooks */
+
+  const loadRemote = React.useCallback(
+    async () => {
+      try {
+        setResult({status: Status.Loading});
+        const importResult = await load(source, getImport, nonce);
+        setResult({status: Status.Complete, imported: importResult});
+      } catch (error) {
+        setResult({status: Status.Failed, error});
+      }
+    },
+    [getImport, nonce, source],
+  );
+
+  React.useEffect(
+    () => {
+      if (
+        result.status === Status.Initial &&
+        defer === DeferTiming.InViewport &&
+        intersection.isIntersecting
+      ) {
+        loadRemote();
+      }
+    },
+    [result, defer, intersection, loadRemote],
+  );
+
+  React.useEffect(
+    () => {
+      if (defer === DeferTiming.Idle) {
+        if ('requestIdleCallback' in window) {
+          idleCallbackHandle.current = (window as WindowWithRequestIdleCallback).requestIdleCallback(
+            loadRemote,
+          );
+        } else {
+          loadRemote();
+        }
+      } else if (defer === DeferTiming.Mount) {
+        loadRemote();
+      }
+
+      return () => {
+        if (
+          idleCallbackHandle.current != null &&
+          typeof (window as any).cancelIdleCallback === undefined
+        ) {
+          (window as any).cancelIdleCallback(idleCallbackHandle.current);
+          idleCallbackHandle.current = null;
+        }
+      };
+    },
+    [defer, loadRemote, intersection, nonce, getImport, source],
```

Comment:
We should make a note in the README that you really can't let any of these values change part-way through rendering, because any change will cancel the load (unlike before)